### PR TITLE
Added second ICA and ica-label functionality

### LINKED ISSUE #5: Add 2nd ICA w/ IClabel to pipeline.
creating tickets for the major remaining TODO's, to incorporate into a 0.1 milestone. 

Description is self explanatory. Maybe we can add parameters to `run_ica` to specify that IC_label should be run, after this line: https://github.com/lina-usc/pylossless/blob/c444b0af2d7bb502e23d82d10dde095d540c71ae/pylossless/pipeline.py#L664

@Andesha made an interesting suggestion, that since the first ICA will always only be used to flag outlying IC timecourses, we may be able to default to using FastICA, with the option to use a different algorithm for the final ICA (if it is more preferable, results in better decompositions).

### PR DESCRIPTION
**Pull request by scott-huberty**
Closes #5 !
I amended the default YAML config file, and `run_ica` to allow for differences in ica_kwargs between run1 and run2.
By default, ica1 will use `fastica` for efficiency, since the purpose of ica1 is only to flag outlying ic timecourses.
By default, ica2 will use `extended infomax`, as this is the default that `mne-icalabel` recommends.

 I also added ica-label funcionality, ic_labels are now stored in `self.ic_labels`.
 I request a review from @christian-oreilly !

### CODE REVIEW

**Comment by scott-huberty** [comment]
I added one more commit here because it relates to the ICA: [61d9d6f](https://github.com/lina-usc/pylossless/pull/7/commits/61d9d6f5d1c38a6621b8c95deb93e53cb0f9e094)

The pipeline will now save the fitted ICA instances (as `.fif` files), and the IC labels (into a .tsv sidecar file).

Screen Shot 2022-10-04 at 10 05 33 AM


 Thus, the EEG visualizer will be able to load in the saved ICA and labels and plot them - So this is a timely addition since we are almost ready to start working on those plots. 

I am going save a lossless derivative into the eeg-visualizer repo so that we can access the raw, ICA, IC labels for the dash development.

File: pylossless/pipeline.py [hunk 1]
```diff
@@ -622,8 +633,34 @@ def flag_epoch_ic_sd1(self, raw):
...
+                                                   )
+        mne_bids.write_raw_bids(raw,
+                                derivatives_path,
+                                overwrite=True,
+                                format='EDF',
+                                allow_preload=True)
+                                #  TODO address derivatives support in MNE bids.
+
+        # Save ICAs
+        for this_ica, self_ica, in zip(['ica1', 'ica2'],
```
Comment (scott-huberty) [inline pylossless/pipeline.py:650]: Maybe we only need to save the final ICA.

Comment (christian-oreilly) [reply pylossless/pipeline.py:650]: Can you remind me of the purpose of ICA1 vs ICA2? Actually, what about documenting this in the default config file? I cannot judge on this question without having a clear idea of its purpose, but I see no reason not to save it. It is a small file and if it is used in the preprocessing, then for reproducibility, it is better to save it.

File: pylossless/pipeline.py [hunk 2]
```diff
@@ -592,24 +595,31 @@ def flag_epoch_gap(self, raw):
...
-        self.ica.fit(epochs)
+        if run == 'run1':
+            self.ica1 = ICA(**ica_kwargs)
+            self.ica1.fit(epochs)
+
+        elif run == 'run2':
+            self.ica2 = ICA(**ica_kwargs)
+            self.ica2.fit(epochs)
+            self.ic_labels = label_components(epochs, self.ica2,
+                                              method="iclabel")
```
Comment (christian-oreilly) [inline pylossless/pipeline.py:616]: Always a good practice to include an "else" statement. In that case, I'd say something like 
```python
else:
     raise ValueError("The `run` arguments can only take 'run1' or 'run2' as value.")
```

**Review by christian-oreilly — CHANGES_REQUESTED** [review]

File: pylossless/pipeline.py [hunk 3]
```diff
@@ -592,24 +595,31 @@ def flag_epoch_gap(self, raw):
...
         raw.set_annotations(raw.annotations + annots)
 
-    def run_ica(self, raw):
-        ica_kwargs = self.config['ica']['ica_args']
+    def run_ica(self, raw, run):
+        ica_kwargs = self.config['ica']['ica_args'][run]
         if 'max_iter' not in ica_kwargs:
             ica_kwargs['max_iter'] = 'auto'
         if 'random_state' not in ica_kwargs:
             ica_kwargs['random_state'] = 97
```
[the remark this answers is not in the document]
Comment (scott-huberty) [reply pylossless/pipeline.py:604]: We coded that together, and yes I think that was just for reproducibility during dev

File: pylossless/pipeline.py (hunk 1 above)
Comment (scott-huberty) [reply pylossless/pipeline.py:650]: the **timecourses** from the ICA1 decomposition are used to flag periods of time that have outlying ICA activation, the same way the pipeline flags bad periods of time with the channel data.

ICA2 is run excluding these bad times (flagged after the prior ICA), and the IC label function is run on this decomposition. This is also the final decomposition used for QCR.

File: pylossless/pipeline.py (hunk 2 above)
Comment (scott-huberty) [reply pylossless/pipeline.py:616]: thanks for the tip! will add this.

Comment (scott-huberty) [reply pylossless/pipeline.py:616]: Hey @christian-oreilly I _think_ we have a problem to be addressed in `run_ica`. If we want to omit bad time segments prior to fitting the ICA (which I think we do) then we need to pass a raw instance into the `fit` method: see the `reject_by_annotation` parameter description in the `ica.fit` method here: https://mne.tools/stable/generated/mne.preprocessing.ICA.html#mne.preprocessing.ICA.fit

Comment (christian-oreilly) [reply pylossless/pipeline.py:618]: @scott-huberty Sorry for not responding to that earlier. If you look at how we coded that, get_epochs(...) returns epochs that already reject any segments with bad data. I think this is OK, unless you are seeing something I don't.

**Review by christian-oreilly — APPROVED** [review]

**Pull request merged**